src/idl_parser.cpp

```diff
@@ -3262,23 +3264,48 @@ CheckedError Parser::ParseRoot(const char *source, const char **include_paths,
   // know for sure what the type of these are.
   for (auto it = enums_.vec.begin(); it != enums_.vec.end(); ++it) {
     auto &enum_def = **it;
+    auto err = CheckPrivateLeak(enum_def);
+    if (err.Check()) { return err; }
     if (enum_def.is_union) {
       for (auto val_it = enum_def.Vals().begin();
            val_it != enum_def.Vals().end(); ++val_it) {
         auto &val = **val_it;
+        if (val.union_type.struct_def) {
+          auto err = CheckPrivateLeak(*val.union_type.struct_def);
+          if (err.Check()) { return err; }
+        }
+
         if (!(opts.lang_to_generate != 0 && SupportsAdvancedUnionFeatures()) &&
             (IsStruct(val.union_type) || IsString(val.union_type)))
+
           return Error(
               "only tables can be union elements in the generated language: " +
               val.name);
       }
     }
   }
+
   // Parse JSON object only if the scheme has been parsed.
   if (token_ == '{') { ECHECK(DoParseJson()); }
   return NoError();
 }
 
+CheckedError Parser::CheckPrivateLeak(const Definition &def) {
+  if (!opts.no_leak_private_annotations) return NoError();
+
+  const auto is_private = def.attributes.Lookup("private");
+  if (module_annotation_ == kNone) {
```

Comment:
I would expect CheckPrivateLeak to be a const function. Its behavior appears to depend on the first definition this parser saw... Can you help me understand what is happening and why?

Reply:
From what I understood is that we wanted to force the compiler to either generate private or public and nothing in between.

Reply:
I think leaks happen when there is a table/struct/union that is not private but it has a field/variant that is private type

Reply:
so forcing all types to be public or private would work but it'll be overkill

Reply:
Definitely, then I will take the weekend to actually focus on this.